packages/cardpay-sdk/sdk/safe-module.ts

```diff
@@ -0,0 +1,417 @@
+import { getAddress } from '../contracts/addresses';
+import {
+  encodeMultiSend,
+  encodeMultiSendCallOnly,
+  getModuleProxyCreationEvent,
+  deployAndSetUpModule,
+  SetupArgs,
+} from './utils/module-utils';
+import {
+  generateSaltNonce,
+  isTransactionHash,
+  Transaction,
+  TransactionOptions,
+  waitUntilTransactionMined,
+} from './utils/general-utils';
+
+import { ContractOptions } from 'web3-eth-contract';
+import {
+  executeTransaction,
+  gasEstimate,
+  generateCreate2SafeTx,
+  getNextNonceFromEstimate,
+  getSafeProxyCreationEvent,
+  Operation,
+} from './utils/safe-utils';
+import { BigNumber, Contract, ContractInterface, Signer, utils } from 'ethers';
+import { signSafeTx, signSafeTxAsBytes } from './utils/signing-utils';
+import { AddressKeys, ERC20ABI } from '..';
+/* eslint-disable node/no-extraneous-import */
+import { AddressZero } from '@ethersproject/constants';
+import BN from 'bn.js';
+import JsonRpcProvider from '../providers/json-rpc-provider';
+import { camelCase } from 'lodash';
+
+export interface EnableModuleAndGuardResult {
+  moduleAddress: string;
+  metaGuardAddress: string;
+}
+
+export interface CreateSafeWithModuleAndGuardResult {
+  safeAddress: string;
+  moduleAddress: string;
+  metaGuardAddress: string;
+}
+
+export interface CreateSafeWithModuleAndGuardTx {
+  multiSendCallOnlyTx: Transaction;
+  expectedSafeAddress: string;
+  expectedModuleAddress: string;
+  expectedMetaGuardAddress: string;
+}
+
+type GasRange = Record<'slow' | 'standard' | 'fast', BigNumber>;
+export interface GasEstimationResult {
+  gas: BigNumber;
+  gasRangeInWei: GasRange;
+  gasRangeInUSD: GasRange;
+}
+
+export const FEE_BASE_POW = new BN(18);
+export const FEE_BASE = new BN(10).pow(FEE_BASE_POW);
+
+export default abstract class SafeModule {
+  public name: AddressKeys;
+  public abstract salt: string;
+  public abstract abi: ContractInterface;
+
+  constructor(protected ethersProvider: JsonRpcProvider, protected signer?: Signer) {
+    this.signer = signer ? signer.connect(ethersProvider) : signer;
+    this.ethersProvider = ethersProvider;
+    this.name = camelCase(this.constructor.name) as AddressKeys;
+  }
+
+  abstract setupArgs(safeAddress: string, safeOwners?: string[]): SetupArgs;
+
+  async createSafeWithModuleAndGuardEstimation(contractOptions?: ContractOptions): Promise<BigNumber> {
+    let signer = this.signer ? this.signer : this.ethersProvider.getSigner();
+    let from = contractOptions?.from ?? (await signer.getAddress());
+
+    let { expectedSafeAddress, create2SafeTx } = await generateCreate2SafeTx(
+      this.ethersProvider,
+      [from],
+      1,
+      AddressZero,
+      '0x',
+      AddressZero,
+      AddressZero,
+      '0',
+      AddressZero,
+      generateSaltNonce(this.salt)
+    );
+    let enableModuleTxs = await this.generateEnableModuleTxs(expectedSafeAddress, [from]);
+    let setGuardTxs = await this.generateSetGuardTxs(expectedSafeAddress);
```

Comment:
Do you need a meta guard for this module?

Reply:
No. I don't. I can refactor it out. But I added it because I copied most of the code from payment scheduler It is causing the multisend to fail. Do you have any idea why the transaction failed 

If you see this address 0xE71F7445A69a2D09d43dE99BE72be79a2C0BD94b 
there are failures here https://goerli.etherscan.io/address/0xE71F7445A69a2D09d43dE99BE72be79a2C0BD94b#internaltx

Reply:
Ive added a function interface that doesn't add metaguard here https://github.com/cardstack/cardstack/pull/3603/commits/74de51cb6fa252f7860325fee33a6b7f3663e9a8

Reply:
Here is the succesful [transaction](https://goerli.etherscan.io/tx/0xe707b0149839d7881c5085f565f440d14e0a09e5171c9950c0d404bc59d25f99)